### ISSUE TITLE
Using python to create BioSQL databases.

### ISSUE THREAD

**Issue by grabear** [biosql/biosql#12]
Would there be any support for creating a Python module that creates template-BioSQL databases?

Is this technology too old?  Or is there a better way to do this already?

If people are interested, we could...

1.  Convert the Perl scripts to Python and make them part of the module.
3.  Update the PhyloDB scripts and schema.
4.  Give users better documentation.

**Comment by grabear** [comment]
https://github.com/ctSkennerton/BioSQL-Extensions @ctSkennerton

**Comment by ctSkennerton** [comment]
I'm not sure what you mean by "template". You can make a blank database without needing any of the perl scripts as long as you don't dump the entire NCBI taxonomy at once but instead build it up as you add in new sequences (as I do). There is already very good support for biosql in the biopython package. @grabear the scripts that you link to are just a simple command line interface to the biopython code.

The PhyloDB tables aren't supported in python at the moment, but I'm not sure if they were ever really completed/integrated in perl either. If this is something that's crucial for you then adding that into biopython would be the best place for it.

**Comment by grabear** [comment]
**To my knowledge:**

There's not a python API for creating a biosql database with the desired schema.  There's also not a python script for uploading NCBI or ITIS taxonomy data into the database.  So I was suggesting that we try to incorporate that here.

I'm creating a module to automate that.  I haven't converted the load_ncbi/itis_taxonomy.pl file to a python script, but I have used subprocess to call this script.  [Here](https://github.com/datasnakes/Datasnakes-Scripts/blob/database-patch/Datasnakes/Manager/database_management.py) is what I'm currently working on to automate the creation of BioSQL databases.  The package as a whole is a bit of a mess, but it's also unfinished.

I say all this in ignorance @ctSkennerton .  You know more than I do, so please let me know if I'm missing something here.  Whenever I first started trying to use biosql databases I was extremely confused, and I think this would help other people who are knew to BioPython, coding, or SQL.

I would also like to update the documentation so that it looks proper on GitHub.

**Comment by grabear** [comment]
Sorry I gave you the wrong link..

https://github.com/datasnakes/Datasnakes-Scripts/blob/database-patch/Datasnakes/Manager/BioSQL/biosql.py

**Comment by ctSkennerton** [comment]
You're right all of the APIs assume that you have a blank database ready to go. A python version of `load_ncbi_taxonomy.pl` would be good. When I looked at it I was very confused at how it worked.

I've never actually looked at the fungal taxonomy script, though doing so just now I realise that it requires the phyloDB extension. Translating those tables to the other database drivers might be a fairly easy first step. I'm not a maintainer of biosql, I just use it quite a bit for my research, but I think if you contribute a pull request it would be appreciated.

**Comment by grabear** [comment]
Ok.  I'll have to work that out, thanks!  I just wanted some feedback from the community, to see if would even be worth my time.  @ctSkennerton  Do you know anything about the PhyloDB extension and the load_itis_taxonomy.pl script?  I haven't used it yet for my tree data yet (in Newick), but I was going to fairly soon.

**Comment by ctSkennerton** [comment]
I've looked through the tables before, they seem to generalize the data structure used for the NCBI taxonomy tables to allow you to store any kind of tree data. I haven't really read the `load_itis_taxonomy.pl` script but it looks like it stores an alternative taxonomy in one of these new tables, since the original schema only had room for NCBIs taxonomy.

**Comment by hlapp** [comment]
There used to be a Biopython API, and I thought there still is – @peterjc do you happen to know the current status of that part in Biopython? Nonetheless, as @ctSkennerton says that API would assume a BioSQL database already in existence.

@grabear if you found something that's missing in Python and are interested in creating it, that'd be awesome, and by all means don't be deterred.

**Comment by peterjc** [comment]
The ``BioSQL`` module in Biopython is still maintained. This includes the capability to incremental update missing entries in the NCBI taxonomy on demand via NCBI Entrez when loading records into the database.

http://biopython.org/wiki/BioSQL
https://github.com/biopython/biopython/tree/master/BioSQL

This code is still maintained (indeed there is an active pull request https://github.com/biopython/biopython/pull/1420 to update our unit tests to be able to run with lower privileges on PosgreSQL), thank you @ctSkennerton who has tested it.

I saw no point re-implementing the perl script included with BioSQL for loading the full taxonomy.

Likewise with MySQL or PostgreSQL so much of setting up a new database was command line based, I saw no point adding anything on top of what the BioSQL project provides.

Biopython does create new databases as part of its test suite by executing bundled copies of the schema SQL files. There would be some benefit in exposing this functionality via the public API under the ``BioSQL`` namespace, you are welcome to explore this and submit a pull request. I think the user would have to provide the path to the BioSQL schema they wish to use (otherwise we have a dependency problem to worry about - or we bundle the latest release of the BioSQL schema with Biopython which is not ideal).

Suggestions or pull requests for improving the BioSQL and/or Biopython documentation are also welcome.

If I have missed any specific questions, please let me know.

**Comment by grabear** [comment]
@peterjc @hlapp @ctSkennerton I know that the BioSQL setup is a lot of stuff to work with (around 5 to 7 MB of data including .png, .pdf, etc..), which is a good argument for not adding to BioPython.  If you exclude the excess files in the setup.py script there will obviously be way less data to import.  If we excluded the oracle schema, derby schema, and hsql schema that would also save some space.

That being said...  
What if instead of adding the scripts to BioPython, we just went ahead and made BioSQL a seperate package on PyPi?  We could create modules with well defined classes and documentation. 
 We may need to change the name though to something different...
maybe BioRDBMS?  

```python
pip install biopython
pip install biordbms
```

The biordbms package could use biopython as a dependency.  So BioPython won't _need_ the biosql package, but it will be necessary for biosql to function with biopython.

**Comment by grabear** [comment]
@sdhutchins This is what I was talking about.  :+1:

**Comment by grabear** [comment]
@azneto

**Comment by grabear** [comment]
We could also just completely abstract the BioSQL module from the BioPython package, and maintain it seperately.

**Comment by peterjc** [comment]
The ``BioSQL`` module shipped with Biopython subclasses the Biopython core objects, meaning releasing it separately on PyPI has implications for careful version compatibility. i.e. It would potentially be more work to maintain. Note Debian splits Biopython into sub-packages, one for the ``Bio`` core and one for ``BioSQL``, but they share version numbers.

I don't yet understand what the benefit of your ``biordbms`` on PyPI idea would be (and what that package would include).

**Comment by grabear** [comment]
If we keep the BioSQL module in the BioPython package, and we create the BioRDBMS (or whatever) package, then we can streamline the process of creating a BioSQL database.  Coming from a non-sql, and non bioinformatic background, I was perplexed by how to do this, initially.  Essentially we can make this more user friendly.  What's the point of having a BioSQL module if only a subset of people can make a BioSQL database?  How many people who are new to BioSQL or bioinformatics were deterred because it was too difficult for their skill level or intelligence?

If we rewrite the various perl scripts in a python module, it will stand along side BioPython in a much more official way.  It would become much easier to contribute extension modules, so that we could make the PhyloDB more tangible.  And so that we could even begin to work on other extensions like Multiple Sequence Alignments, or even RNA-seek data.

It would also all of a sudden be much easier to document.

**Comment by peterjc** [comment]
Personally I see no point rewriting the existing BioSQL Perl scripts in Python, just to put the BioSQL schemes and scripts into PyPI (Python's package system, the equivalent of CPAN for Perl, etc). BioSQL is meant to be language neutral with wrappers in Perl, Python, Ruby and Java (in the Bio* projects).

I think a better idea would be putting together a BioConda package for BioSQL, which should make it easy to install and then use with any of the language specific bindings. Conda is language neutral, and the BioConda channel has a vast number of bioinformatics tools packaged up for one-line installation or update.
